internal/civisibility/integrations/gotesting/instrumentation.go

```diff
@@ -128,492 +128,344 @@ func deleteTestMetadata(tb testing.TB) {
+
+	// Module and suite for this test
+	var module integrations.DdTestModule
+	var suite integrations.DdTestSuite
+
+	// Check if we have execution metadata to propagate
+	originalExecMeta := getTestMetadata(t)
+
+	retryCount := initialRetryCount
+
+	for {
+		// Clear the matcher subnames map before each execution to avoid subname tests being called "parent/subname#NN" due to retries
+		getTestContextMatcherPrivateFields(t).ClearSubNames()
+
+		// Increment execution index
+		executionIndex++
+
+		// Create a new local copy of `t` to isolate execution results
+		ptrToLocalT := &testing.T{}
+		copyTestWithoutParent(t, ptrToLocalT)
+
+		// Create a dummy parent so we can run the test using this local copy
+		// without affecting the test parent
+		localTPrivateFields := getTestPrivateFields(ptrToLocalT)
+		*localTPrivateFields.parent = unsafe.Pointer(&testing.T{})
+
+		// Create an execution metadata instance
+		execMeta := createTestMetadata(ptrToLocalT)
+		execMeta.hasAdditionalFeatureWrapper = true
+
+		// Propagate set tags from a parent wrapper
+		if originalExecMeta != nil {
+			if originalExecMeta.isANewTest {
+				execMeta.isANewTest = true
+			}
+			if originalExecMeta.isARetry {
+				execMeta.isARetry = true
+			}
 		}
-		// Increment the test count in the module.
-		atomic.AddInt32(modulesCounters[moduleName], 1)
 
-		// Initialize suite counters if not already present.
-		if _, ok := suitesCounters[suiteName]; !ok {
-			var v int32
-			suitesCounters[suiteName] = &v
+		// If we are in a retry execution, set the `isARetry` flag
+		if executionIndex > 0 {
+			execMeta.isARetry = true
 		}
-		// Increment the test count in the suite.
-		atomic.AddInt32(suitesCounters[suiteName], 1)
 
-		// Decrement level.
-		bpf := getBenchmarkPrivateFields(b)
-		bpf.AddLevel(-1)
+		// Adjust execution metadata
+		if execMetaAdjust != nil {
+			execMetaAdjust(execMeta, executionIndex)
+		}
 
+		// Run original func similar to how it gets run internally in tRunner
 		startTime := time.Now()
-		module := session.GetOrCreateModuleWithFrameworkAndStartTime(moduleName, testFramework, runtime.Version(), startTime)
-		suite := module.GetOrCreateSuiteWithStartTime(suiteName, startTime)
-		test := suite.CreateTestWithStartTime(fmt.Sprintf("%s/%s", pb.Name(), name), startTime)
-		test.SetTestFunc(originalFunc)
-
-		// Restore the original name without the sub-benchmark auto name.
-		*bpf.name = subBenchmarkAutoNameRegex.ReplaceAllString(*bpf.name, "")
-
-		// Run original benchmark.
-		var iPfOfB *benchmarkPrivateFields
-		var recoverFunc *func(r any)
-		instrumentedFunc := func(b *testing.B) {
-			// Stop the timer to do the initialization and replacements.
-			b.StopTimer()
-
+		chn := make(chan struct{}, 1)
+		go func() {
 			defer func() {
-				if r := recover(); r != nil {
-					if recoverFunc != nil {
-						fn := *recoverFunc
-						fn(r)
-					}
-					panic(r)
-				}
+				chn <- struct{}{}
 			}()
+			targetFunc(ptrToLocalT)
+		}()
+		<-chn
```

Comment:
What is the difference between starting a goroutine and waiting for it to finish vs. just calling the function in the current goroutine? I'm not familiar with this.

Reply:
the internal go testing.tRunner do the same for a very good reason. The function `t.SkipNow()` and `t.FailNow()` calls `runtime.Goexit()` and this function do:
```
// Goexit terminates the goroutine that calls it. No other goroutine is affected.
// Goexit runs all deferred calls before terminating the goroutine. Because Goexit
// is not a panic, any recover calls in those deferred functions will return nil.
```
So we are kind of forced to use goroutines because we don't want the retry controller to be terminated by this function. The nice thing is that all deferred calls gets called so the channel will work.